demo/cts_poc/notary-sign.sh

```diff
@@ -0,0 +1,113 @@
+: "${SIGNATURE_OUTPUT_PATH:?"variable not set. Please define the path where the COSE signature will be saved to"}"
+
+# Optional variables
+IMAGE_TAG=${IMAGE_TAG:-""}
+IMAGE_DIGEST=${IMAGE_DIGEST:-""}
+
+# Optional variables for Notation signing
+# Specify true or false whether the certificate in AKV is a self-signed one or not
+IS_SELF_SIGNED_CERT=${IS_SELF_SIGNED_CERT:-""}
+# Specify the path to the CA certificates PEM file if the certificate in AKV does not contain the full certificate chain
+CA_CERTS_PEM_FILE_PATH=${CA_CERTS_PEM_FILE_PATH:-""}
+
+# Check if the image tag or digest is provided
+if [ -z "$IMAGE_TAG" ] && [ -z "$IMAGE_DIGEST" ]; then
+    echo "Either IMAGE_TAG or IMAGE_DIGEST must be provided"
+    exit 1
+fi
+
+echo -e "\nInstall ORAS CLI"
+
+# Install ORAS CLI
+ORAS_VERSION="1.1.0"
+curl -LO "https://github.com/oras-project/oras/releases/download/v${ORAS_VERSION}/oras_${ORAS_VERSION}_linux_amd64.tar.gz"
+mkdir -p oras-install/
+tar -zxf oras_${ORAS_VERSION}_*.tar.gz -C oras-install/
+mv oras-install/oras /usr/local/bin/
+rm -rf oras_${ORAS_VERSION}_*.tar.gz oras-install/
+
+echo -e "\nInstall Notation CLI"
+
+# Download, extract, and install notation
+NOTATION_VERSION="1.0.1"
+curl -Lo notation.tar.gz https://github.com/notaryproject/notation/releases/download/v${NOTATION_VERSION}/notation_${NOTATION_VERSION}_linux_amd64.tar.gz
+mkdir -p ~/temp_notation_dir
+tar xzf notation.tar.gz -C ~/temp_notation_dir
+
+# Copy the Notation binary to the desired bin directory in your $PATH, for example
+mv ~/temp_notation_dir/notation /usr/local/bin
+rm -rf notation.tar.gz ~/temp_notation_dir
+
+echo -e "\nInstall Notation Azure Key Vault plugin"
+
+# Create a directory for the plugin
+mkdir -p ~/.config/notation/plugins/azure-kv
+
+# Download the plugin
+curl -Lo notation-azure-kv.tar.gz \
+    https://github.com/Azure/notation-azure-kv/releases/download/v${NOTATION_VERSION}/notation-azure-kv_${NOTATION_VERSION}_linux_amd64.tar.gz 
+
+# Extract to the plugin directory
+tar xzf notation-azure-kv.tar.gz -C ~/.config/notation/plugins/azure-kv
+rm -rf notation-azure-kv.tar.gz
+
+echo -e "\nSign container image with Notation"
+
+# Get the KID of the certificate in Azure Key Vault
+KEY_ID=$(az keyvault certificate show -n "$CERTIFICATE_NAME" --vault-name "$AKV_NAME" --version "$CERTIFICATE_VERSION" --query 'kid' -o tsv)
+
+# Login to ACR
+az acr login --name "$ACR_NAME"
+
+REGISTRY="$ACR_NAME".azurecr.io
+
+# Concatenate the image repository to either the image tag or digest
+if [ -n "$IMAGE_DIGEST" ]; then
+    TBS_IMAGE="$IMAGE_REPOSITORY@$IMAGE_DIGEST"
+else
+    TBS_IMAGE="$IMAGE_REPOSITORY:$IMAGE_TAG"
+fi
+
+IMAGE=$REGISTRY/$TBS_IMAGE
+
+# Construct additional notation sign arguments based on environment variable values
+NOTATION_AKV_PLUGIN_ARGS=""
+if [ "$IS_SELF_SIGNED_CERT" = "true" ]; then
+    NOTATION_AKV_PLUGIN_ARGS="--plugin-config self_signed=true"
+elif [ -n "$CA_CERTS_PEM_FILE_PATH" ]; then
```

Comment:
notation is a great example here to show that the certs are loaded from the directory before adding them to the x5c header

Reply:
It's actually not the case. The `CA_CERTS_PEM_FILE_PATH` is an optional parameter that can be passed to notation in case the root and intermediate CA certs are not already included in the certificate in AKV. If the certificate in AKV already includes the full chain, this parameter should not be passed.

The AKV notation plugin will automatically sign the container image using the certificate in KV and attach the certificate chain to the x5c header without further actions.

Reply:
Indeed, looks like the plugin is using the same mechanism (fetching the secret) to get the full certificate chain: https://github.com/Azure/notation-azure-kv/blob/3d8f4af55238560cd89b57fc862561b185858768/Notation.Plugin.AzureKeyVault/KeyVault/KeyVaultClient.cs#L186  😄

Reply:
Yeah then we should not do it, I did not know they are fetching the key. It is forcing the user to have access to the private key